algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala

```diff
@@ -0,0 +1,405 @@
+ * structure with
+ *
+ * - epsilon, the relative error you're willing to tolerate
+ * - windowSize, the number of time ticks that you want to track
+ *
+ * You interact with the data structure by adding (number, timestamp)
+ * pairs into the exponential histogram. querying it for an
+ * approximate counts with `guess`.
+ *
+ * The approximate count is guaranteed to be within conf.epsilon
+ * relative error of the true count seen across the supplied
+ * `windowSize`.
+ *
+ * Next steps:
+ *
+ * - efficient serialization
+ * - Query EH with a shorter window than the configured window
+ * - Discussion of epsilon vs memory tradeoffs
+ *
+ *
+ * @param conf the config values for this instance.
+ * @param buckets Vector of timestamps of each (powers of 2)
+ *                ticks. This is the key to the exponential histogram
+ *                representation. See [[ExpHist.Canonical]] for more
+ *                info.
+ * @param total total ticks tracked. `total == buckets.map(_.size).sum`
+ * @param time current timestamp of this instance.
+ */
+case class ExpHist(conf: ExpHist.Config, buckets: Vector[ExpHist.Bucket], total: Long, time: Long) {
+  import ExpHist.{ Bucket, Canonical }
+
+  /**
+   * Steps this instance forward to the new supplied time. Any
+   * buckets with a timestamp <= (newTime - conf.windowSize) will be
+   * evicted.
+   *
+   * @param newTime the new current time.
+   * @return ExpHist instance stepped forward to newTime.
+   */
+  def step(newTime: Long): ExpHist =
+    if (newTime <= time) this
+    else {
+      val (dropped, filtered) = conf.dropExpired(buckets, newTime)
+      copy(time = newTime, buckets = filtered, total = total - dropped)
+    }
+
+  /**
+   * Increment ExpHist by 1 at the supplied timestamp.
+   */
+  def inc(timestamp: Long): ExpHist = add(1L, timestamp)
+
+  /**
+   * Increment ExpHist by delta at the supplied timestamp.
+   */
+  def add(delta: Long, timestamp: Long): ExpHist = {
+    val self = step(timestamp)
+    if (delta == 0) self
+    else self.addAllWithoutStep(Vector(Bucket(delta, timestamp)), delta)
+  }
+
+  /**
+   * Efficiently add many buckets at once.
+   *
+   * @param unsorted vector of buckets. All timestamps must be >= this.time.
+   * @return ExpHist instance with all buckets added, stepped
+   *         forward to the max timestamp in `unsorted`.
+   */
+  def addAll(unsorted: Vector[Bucket]): ExpHist =
+    if (unsorted.isEmpty) this
+    else {
+      val sorted = unsorted.sorted(Ordering[Bucket].reverse)
+      val delta = sorted.map(_.size).sum
+      val timestamp = sorted.head.timestamp
+      if (delta == 0)
+        step(timestamp)
+      else {
+        addAllWithoutStep(sorted, delta).step(timestamp)
+      }
+    }
+
+  /**
+   * Returns a [[Fold]] instance that uses `add` to accumulate deltas
+   * into this exponential histogram instance.
+   */
+  def fold: Fold[Bucket, ExpHist] =
+    Fold.foldLeft(this) {
```

Comment:
I wonder about using a `foldMutable` here and accumulating a batch of `Bucket` and then doing an `addAll` which might be a lot faster, but we can punt on this if you like.